lib/hsm.js

```diff
@@ -0,0 +1,236 @@
+/* eslint no-case-declarations: "off" */
+
+/*!
+ * hsm.js - HSM PKCS#11 interface for bns
+ * Copyright (c) 2021, Matthew Zipkin (MIT License).
+ * https://github.com/chjj/bns
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const pkcs11js = require('pkcs11js');
+const constants = require('./constants');
+const dnssec = require('./dnssec');
+const util = require('./util');
+const {Record, DNSKEYRecord, RRSIGRecord, types} = require('./wire');
+
+const {algs} = constants;
+
+// Comment from pkcs11js pkcs11t.h:
+// CKK_ECDSA is deprecated in v2.11, CKK_EC is preferred.
+const algToKeyType = {
+  [algs.RSASHA1]: pkcs11js.CKK_RSA,
+  [algs.RSASHA256]: pkcs11js.CKK_RSA,
+  [algs.ECDSAP256SHA256]: pkcs11js.CKK_EC,
+  [algs.RSASHA512]: pkcs11js.CKK_RSA
+};
+
+const algToMechanism = {
+  [algs.RSASHA1]: pkcs11js.CKM_RSA_PKCS,
+  [algs.RSASHA256]: pkcs11js.CKM_RSA_PKCS,
+  [algs.ECDSAP256SHA256]: pkcs11js.CKM_ECDSA,
+  [algs.RSASHA512]: pkcs11js.CKM_RSA_PKCS
+};
+
+class HSMUser {
+  constructor(options) {
+    this.module = null;
+    this.pkcs11 = null;
+    this.session = null;
+    this.slot = Buffer.alloc(8);
+    this.pin = null;
+
+    if (options)
+      this.fromOptions(options);
+  }
+
+  fromOptions(options) {
+    if (options.module != null) {
+      assert(typeof options.module === 'string');
+      this.module = options.module;
+    }
+
+    if (options.slot != null) {
+      assert(Number.isSafeInteger(options.slot));
+      assert(options.slot >= 0);
+      assert(options.slot <= 0xffffffff);
+      this.slot.writeUInt32LE(options.slot);
+    }
+
+    if (options.pin != null) {
+      assert(typeof options.pin === 'string');
+      this.pin = options.pin;
+    }
+  }
+
+  open() {
+    // Load module
+    this.pkcs11 = new pkcs11js.PKCS11();
+    this.pkcs11.load(this.module);
+
+    // Initialize library
+    this.pkcs11.C_Initialize();
+
+    // Create read-only session
+    // by not setting pkcs11js.CKF_RW_SESSION flag
+    this.session = this.pkcs11.C_OpenSession(
+      this.slot,
+      pkcs11js.CKF_SERIAL_SESSION
+    );
+
+    // Login in user
+    this.pkcs11.C_Login(this.session, pkcs11js.CKU_USER, this.pin);
+  }
+
+  sign(dnskey, rrset) {
+    assert(dnskey instanceof Record);
+    assert(dnskey.data instanceof DNSKEYRecord);
+    assert(util.isRRSet(rrset));
+
+    // Use DNSKEY record to find the private key we need to sign with
+    const [alg, pub] = this.extractPublicKey(dnskey);
+    const priv = this.getPrivForPub(alg, pub);
+
+    // Prepare data for signing
+    const sig = new Record();
+    const rd = new RRSIGRecord();
+    sig.name = dnskey.name;
+    sig.ttl = dnskey.ttl;
+    sig.class = dnskey.class;
+    sig.type = types.RRSIG;
+    sig.data = rd;
+    rd.keyTag = dnskey.data.keyTag();
+    rd.signerName = dnskey.name;
+    rd.algorithm = alg;
+    rd.inception = util.now() - (24 * 60 * 60);
```

Comment:
Any specific reason to have the inception time a day before?

Reply:
This is how we do it in bns `dnssec.sign()` as well: https://github.com/chjj/bns/blob/master/lib/dnssec.js#L191

My guess is, its just to ensure that when you deploy there will be very few machines on earth that think the timestamp is invalid. This is especially true for blockchains which have a weird concept of "now".